src/MongoDB.EntityFrameworkCore/Storage/MongoTypeMappingSource.cs

```diff
@@ -76,27 +83,46 @@ public MongoTypeMappingSource(TypeMappingSourceDependencies dependencies)
             var elementMapping = FindPrimitiveMapping(elementMappingInfo) ?? FindCollectionMapping(elementMappingInfo);
             return elementMapping == null
                 ? null
-                : new MongoTypeMapping(clrType, ComparerFactory.CreateComparer(elementMapping, clrType));
+                : new MongoTypeMapping(clrType, CreateComparer(elementMapping, clrType));
         }
 
         // Support collection on the entity
-        if (clrType.IsGenericType && !clrType.IsGenericTypeDefinition)
+        if (clrType is {IsGenericType: true, IsGenericTypeDefinition: false})
         {
             var genericTypeDefinition = clrType.GetGenericTypeDefinition();
             if (genericTypeDefinition == typeof(List<>)
                 || genericTypeDefinition == typeof(IList<>)
-                || genericTypeDefinition == typeof(ICollection<>)
                 || genericTypeDefinition == typeof(IReadOnlyList<>)
-                || genericTypeDefinition == typeof(IReadOnlyCollection<>))
+                || genericTypeDefinition == typeof(ObservableCollection<>)
+                || genericTypeDefinition == typeof(Collection<>))
             {
                 var elementMappingInfo = new TypeMappingInfo(itemType);
                 var elementMapping = FindPrimitiveMapping(elementMappingInfo) ?? FindCollectionMapping(elementMappingInfo);
                 return elementMapping == null
                     ? null
-                    : new MongoTypeMapping(clrType, ComparerFactory.CreateComparer(elementMapping, clrType));
+                    : new MongoTypeMapping(clrType, CreateComparer(elementMapping, clrType));
             }
         }
 
         return null;
     }
+
+    public static ValueComparer? CreateComparer(CoreTypeMapping elementMapping, Type collectionType)
+    {
+        // Only single dimension arrays are supported
+        if (collectionType.IsArray && collectionType.GetArrayRank() != 1)
+        {
+            return null;
```

Comment:
Should we throw here?

Reply:
This block can now actually be removed as the new value comparers support multi-dimensional arrays.

If we don't support that in the C# Driver/BsonDocument then we should add detection to that as part of the model validation instead where we can throw the right error long before anything gets here.